---
title: "`r params$project_title`"
editor_options:
  chunk_output_type: console
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    toc: yes
  bookdown::html_document2:
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: no
params:
  dataset: 1
---

```{r Setup, echo = FALSE}
knitr::opts_chunk$set(fig.width  = 7,
                      results    = "hold",
                      message    = FALSE,
                      warning    = FALSE,
                      echo = FALSE)


```

```{r setup}

suppressPackageStartupMessages({
    library(Seurat)
    library(venn)
    library(dplyr)
    library(cowplot)
    library(ggplot2)
    library(pheatmap)
    library(enrichR)
    library(rafalib)
    library(cowplot)
  library(gplots) 
  library(ggrepel)
  library(reshape2)})


plottheme = theme(legend.title = element_text(size = 8),
                  legend.text = element_text(size = 8),
                  axis.title = element_text(size = 8),
                  axis.text = element_text(size = 8),
                  plot.title = element_text(size = 8))


# library(future)
# plan("multiprocess", workers = 4)
# options(future.globals.maxSize = 24000 * 1024^2)

alldata <- readRDS("../analysis/filtered_IMM_DN_int_clus.rds")

# Set desired clustering resolution
sel.lev = 0.25

# Set the identity as louvain with resolution selected above
sel.clust = paste0("CCA_snn_res.",sel.lev)

alldata@active.assay = "RNA"

alldata$orig.ident = factor(alldata$orig.ident, 
                            levels = c("STD_DN_d0","GW_DN_d0",
                                       "STD_IMM_d0","GW_IMM_d0"))
alldata$Treatment = factor(alldata$Treatment, levels = c("STD","GW"))

alldata <- SetIdent(alldata, value = sel.clust)

alldata$ident_clust = paste0(alldata@active.ident,"_",alldata$orig.ident)

ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

color_list = ggplotColours(10)[c(2,4,7,9)]

names(color_list) = levels(alldata$orig.ident)

cluster_colors = ggplotColours(length(unique(alldata@active.ident))+1)[-1]

```

# Used clustering

The clustering resolution being analyzed is shown below. The number of cells per cluster are as follows:

`r table(alldata@active.ident)`


```{r clusteringdim}

plot_grid(ncol = 2, DimPlot(alldata, label = T, shuffle = TRUE, raster = TRUE) + 
            NoAxes() + NoLegend() + scale_color_manual(values = cluster_colors),
          DimPlot(alldata, group.by = "orig.ident", shuffle = TRUE, raster = TRUE) + 
            scale_color_manual(values = color_list) +
            NoAxes() + theme(legend.position = "bottom") +
  guides(color=guide_legend(ncol=2,override.aes = list(size = 2))))

```

```{r, fig.width=4, fig.height = 4}

ggplot(alldata@meta.data, aes_string(fill= sel.clust, x = "Treatment")) + plottheme + 
  geom_bar(position = "fill", color = "black") + facet_wrap(~Type) +
  theme(panel.background = element_blank(),axis.ticks = element_blank(), axis.title.y = element_blank(),
        legend.key.width = unit(7,"pt"), legend.key.size = unit(3,"pt"))

```


The differences between proportions of cells from each cluster in each sample can be seen in the following barplots.

```{r, fig.height =3, fig.width=7}

frequencies = as.data.frame(table(data.frame(cluster=alldata@meta.data[,sel.clust],
                 sample = alldata$orig.ident, treatment = alldata$Treatment,
                 type = alldata$Type)))
frequencies$Proportion=NA
for(i in 1:nrow(frequencies)){
  frequencies$Proportion[i] = frequencies$Freq[i]/sum(frequencies$Freq[frequencies$sample==frequencies$sample[i]])
}

ggplot(frequencies, aes(fill=treatment, y = Proportion, x = cluster)) + plottheme + 
  geom_bar(stat = "identity",position="dodge") + facet_wrap(~type) +
  theme(panel.background = element_blank(),axis.ticks.y = element_blank())

```


# Annotating clusters

The clusters are annotated using information from the samples and the gene expression.


```{r, fig.height = 3}

plot_grid(ggplot(frequencies, aes(fill=type, y = Proportion, x = cluster)) + plottheme + 
  geom_bar(stat = "identity", position = "fill") + 
  theme(panel.background = element_blank(),axis.ticks.y = element_blank()),

d = DotPlot(alldata, features = c("Ptprc","Epcam"), 
    assay = "RNA") + coord_flip() + plottheme + theme(legend.position = "bottom"))

```


Top 25 (by p-value) genes with logFC>0.2 for each cluster, sorted by logFC:


```{r,  fig.height = 4}


if(file.exists(paste0("../analysis/clustermarkers_IMM_DN_CCA_",sel.lev,".rds"))){
  markers_genes = readRDS(paste0("../analysis/clustermarkers_IMM_DN_CCA_",sel.lev,".rds"))
}else{
  markers_genes <- FindAllMarkers(alldata, logfc.threshold = 0.2, test.use = "wilcox", 
  min.pct = 0.1, min.diff.pct = 0.2, only.pos = TRUE, max.cells.per.ident = 300, 
  assay = "RNA", random.seed = 42)
  saveRDS(markers_genes, paste0("../analysis/clustermarkers_IMM_DN_CCA_",sel.lev,".rds"))
}


top25 <- markers_genes %>% group_by(cluster) %>% top_n(-20, p_val)
top25 <- top25 %>% group_by(cluster) %>% top_n(20, avg_log2FC)


mypar(1, 5, mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F), horiz = T, 
        las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}



```


Top 5 (by p-value) genes with logFC>0.2 for each cluster:

```{r, fig.height=12}

top5 <- markers_genes %>% group_by(cluster) %>% top_n(-5, p_val_adj)
top5 <- top5 %>% group_by(cluster) %>% top_n(5, avg_log2FC)

alldata <- ScaleData(alldata, features = as.character(unique(top5$gene)), assay = "RNA")
DoHeatmap(alldata, features = as.character(unique(top5$gene)), group.by = sel.clust, 
    assay = "RNA") + plottheme

DotPlot(alldata, features = rev(as.character(unique(top5$gene))), 
    assay = "RNA") + coord_flip() + plottheme

```

Top 3 (by p-value) genes with logFC>0.2 for each cluster:

```{r, fig.height=1.5, fig.width = 1.5}

top3 <- top5 %>% group_by(cluster) %>% top_n(-3, p_val)
top3 <- top3 %>% group_by(cluster) %>% top_n(3, avg_log2FC)

for(g in unique(top3$gene)){
  print(VlnPlot(alldata, features = g, group.by = sel.clust, 
    assay = "RNA", pt.size = -1) + theme(axis.title = element_blank()) +
      NoLegend())
}


```

```{r, fig.height=1.5, fig.width = 1.5}

for(g in as.character(unique(top3$gene))){
  print(FeaturePlot(alldata, features = g, slot = "data", raster = TRUE) +
    NoAxes() + plottheme + theme(legend.key.width = unit(5, "pt"), 
                                 legend.key.height = unit(5, "pt")))
}

```


## Annotation labels

Based on the top DE genes, here are my predctions for the cell types represented by each cluster:

* Cluster 0: T cells
* Cluster 1: Adamdec1+ fibroblasts
* Cluster 2: Plasma cells
* Cluster 3: Dcn+ fibroblasts
* Cluster 4: B cells
* Cluster 5: Trophocytes
* Cluster 6: Myeloid cells
* Cluster 7: ILC3
* Cluster 8: Bmp5+ fibroblasts
* Cluster 9: Myocytes
* Cluster 10: Endothelial
* Cluster 11: Hhip+ nyocytes
* Cluster 12: Lef1+ lymphocytes
* Cluster 13: ILC2
* Cluster 14: Proliferating
* Cluster 15: pDCs
* Cluster 16: LECs

# Diet comparisons

## Full samples

### Immune


```{r, fig.height = 3}

volcanotheme = theme(axis.text = element_text(size = 8),
                     axis.title = element_text(size = 8),
                     panel.grid = element_blank(),
                     panel.background = element_blank(),
                     
                    plot.title = element_text(hjust = 0.5,
                                              size = 10))


data.types = SplitObject(alldata, split.by = "Type")


if(!file.exists("../analysis/imm_de_all.rds")){
imm_de = FindMarkers(data.types[["IMM"]],
            ident.1 = "GW",ident.2 = "STD",
            group.by = "Treatment", logfc.threshold = -Inf)
saveRDS(imm_de, "../analysis/imm_de_all.rds")
}else{
  imm_de = readRDS("../analysis/imm_de_all.rds")
}



imm_de$gene = rownames(imm_de)

vdata = imm_de
g_all = ggplot(vdata, aes(x = avg_log2FC, y = -log10(p_val_adj))) + 
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) + 
  geom_point(size = 1) + 
  geom_text_repel(data = vdata[abs(vdata$avg_log2FC)>1,],
                  aes(label = gene), size = 2.5) + 
  labs(title = "All cells") +
  volcanotheme

imm_de_500 = FindMarkers(data.types[["IMM"]],
            ident.1 = "GW",ident.2 = "STD",
            group.by = "Treatment", logfc.threshold = -Inf,
            max.cells.per.ident = 500)

imm_de_500$gene = rownames(imm_de_500)

vdata = imm_de_500
g_500 = ggplot(vdata, aes(x = avg_log2FC, y = -log10(p_val_adj))) + 
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) + 
  geom_point(size = 1) + 
  geom_text_repel(data = vdata[abs(vdata$avg_log2FC)>1,],
                  aes(label = gene), size = 2.5) + 
  labs(title = "500 cells per group") +
  volcanotheme

plot_grid(g_all,
          g_500)

```

### DN

```{r, fig.height = 3}

data.types = SplitObject(alldata, split.by = "Type")

if(!file.exists("../analysis/dn_de_all.rds")){
dn_de = FindMarkers(data.types[["DN"]],
            ident.1 = "GW",ident.2 = "STD",
            group.by = "Treatment", logfc.threshold = -Inf)
saveRDS(dn_de, "../analysis/dn_de_all.rds")
}else{
  dn_de = readRDS("../analysis/dn_de_all.rds")
}


dn_de$gene = rownames(dn_de)

vdata = dn_de
g_all = ggplot(vdata, aes(x = avg_log2FC, y = -log10(p_val_adj))) + 
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) + 
  geom_point(size = 1) + 
  geom_text_repel(data = vdata[abs(vdata$avg_log2FC)>1,],
                  aes(label = gene), size = 2.5) + 
  labs(title = "All cells") +
  volcanotheme


dn_de_500 = FindMarkers(data.types[["DN"]],
            ident.1 = "GW",ident.2 = "STD",
            group.by = "Treatment", logfc.threshold = -Inf,
            max.cells.per.ident = 500)

dn_de_500$gene = rownames(dn_de_500)

vdata = dn_de_500
g_500 = ggplot(vdata, aes(x = avg_log2FC, y = -log10(p_val_adj))) + 
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) + 
  geom_point(size = 1) + 
  geom_text_repel(data = vdata[abs(vdata$avg_log2FC)>1,],
                  aes(label = gene), size = 2.5) + 
  labs(title = "500 cells per group") +
  volcanotheme

plot_grid(g_all,
          g_500)

```


## Per cluster

### All cells

For each cluster, the differentially expressed genes between the two treatments are identified. The results are shown below: Dot plots for the top 20 DE genes and volcano plots for all DE genes. Note that the higher significance in the first clusters is primarily due to the sizes of the clusters.

```{r de, fig.height = 4, fig.width = 2.5}


if(!file.exists(paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,".rds"))){
DGE_list_Treatment = list()

  for(i in sort(unique(alldata@active.ident))){
  message(paste("Cluster",i))
  # select all cells in cluster i
  cell_selection <- subset(alldata, cells = colnames(alldata)[alldata@active.ident == 
      i])
  cell_selection <- SetIdent(cell_selection, value = "Treatment")
  # Compute differential expression
  DGE_cell_selection <- FindMarkers(ident.1 = "GW", ident.2 = "STD",
                                    cell_selection, logfc.threshold = 0.2,
                                    test.use = "wilcox", 
      min.pct = 0.1, only.pos = FALSE,
      assay = "RNA")
  DGE_list_Treatment[[paste("Cluster",i)]] = DGE_cell_selection[DGE_cell_selection$p_val_adj<0.001,]
  DGE_list_Treatment[[paste("Cluster",i)]]$gene = rownames(DGE_list_Treatment[[paste("Cluster",i)]])
  saveRDS(DGE_list_Treatment, paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,".rds"))
}}else{
  DGE_list_Treatment = readRDS(paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,".rds"))
}

```


```{r}

num.cells = as.matrix(table(alldata@active.ident,
                  alldata$Treatment))
rownames(num.cells) = paste0("Cluster ", rownames(num.cells))

reglist = lapply(names(DGE_list_Treatment), function(i){
  message(i)
  x = DGE_list_Treatment[[i]]
  c(up = sum(x$avg_log2FC > 0),
    down = sum(x$avg_log2FC < 0),
    total = nrow(x),
    min.cells = min(num.cells[i,]))})
names(reglist) = names(DGE_list_Treatment)
regdf = as.data.frame(reglist)
colnames(regdf) = gsub("Cluster.","", colnames(regdf))

```

Number of DE genes per cluster

`r knitr::kable(regdf)`

Below are the numbers of DE genes per cluster according to size of cluster. 

```{r,fig.height = 4, fig.width = 4}


tregdf = as.data.frame(t(regdf))
tregdf$Cluster = rownames(tregdf)
ggplot(tregdf, aes(x = min.cells, y = total)) + 
  geom_text(aes(label = Cluster ))

```

```{r dotplots, fig.height = 4, fig.width = 2.5}
for(i in sort(unique(alldata@active.ident))){
  top20_cell_selection <- DGE_list_Treatment[[paste("Cluster",i)]]
  n = min(nrow(top20_cell_selection), 20)
  top20_cell_selection <- top20_cell_selection[order(top20_cell_selection$p_val)[1:n],]
  top20_cell_selection <- top20_cell_selection[order(top20_cell_selection$avg_log2FC)[1:n],]
  # print(VlnPlot(subset(alldata, idents = i),unique(top20_cell_selection$gene), 
  #         group.by = "orig.ident", assay = "RNA") + theme(axis.text = element_text(size = 1)))
  if(nrow(top20_cell_selection)>0){
    d = DotPlot(subset(alldata, idents = i), scale = FALSE, 
                features = rev(as.character(unique(top20_cell_selection$gene))), 
      assay = "RNA", group.by = "Treatment") + coord_flip() + plottheme +
        labs(title = paste("Cluster (",i,paste(top3$gene[top3$cluster==i],collapse = ", "),
                           ") (",nrow(DGE_list_Treatment[[paste("Cluster",i)]])," genes)"))
    d$guides$size$title = "Percent"
    d$guides$colour$title = "Average"
    print(d)
  }
}

```

```{r volcanos, fig.height = 4, fig.width = 7}


for(i in sort(unique(alldata@active.ident))){
  res = DGE_list_Treatment[[paste("Cluster",i)]]
  
  #res$avg_log2FC[res$cluster=="STD"] = 0-res$avg_log2FC[res$cluster=="STD"]
  
  if(nrow(res)>0){
      print(ggplot(res, aes(x = avg_log2FC, y = -log10(p_val_adj))) +
              geom_point(color = "#AAAAAA") + 
              geom_hline(yintercept = 0) +
              geom_vline(xintercept = 0) + 
              geom_text_repel(aes(label = gene), size = 2.5) +
              plottheme +
              theme(panel.background = element_blank(),
                    panel.grid = element_blank()) + 
        labs(title = paste("Cluster",i,paste(top3$gene[top3$cluster==i],collapse = ", "))))
  
  }
}



```



```{r heatmaps, fig.height = 10, fig.width = 8, include = FALSE}

# allde = unique(do.call("rbind",DGE_list_Treatment)$gene)
# 
# #allde
# 
# if(!file.exists( paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_allde.rds"))){
# DGE_list_Treatment_de = list()
# 
# for(i in sort(unique(alldata@active.ident))){
#   # select all cells in cluster i
#   cell_selection <- subset(alldata, cells = colnames(alldata)[alldata@active.ident ==
#       i])
#   cell_selection <- SetIdent(cell_selection, value = "Treatment")
#   # Compute differentiall expression
#   DGE_cell_selection <- FindMarkers(cell_selection, ident.1 = "GW",
#                                     ident.2 = "STD", features = allde,
#                                     logfc.threshold = -Inf, test.use = "wilcox",
#       min.pct = 0, only.pos = FALSE,
#       assay = "RNA")
#   DGE_list_Treatment_de[[paste("Cluster",i)]] = DGE_cell_selection
#   saveRDS(DGE_list_Treatment_de, paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_allde.rds"))
# }}else{
#   DGE_list_Treatment_de = readRDS(paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_allde.rds"))
# }
# 
# pvals = as.data.frame(lapply( DGE_list_Treatment_de, function(x){x[allde,"p_val_adj"]}))
# rownames(pvals) <- allde
# 
# logfcs = as.data.frame(lapply( DGE_list_Treatment_de, function(x){x[allde,"avg_log2FC"]}))
# rownames(logfcs) <- allde
# 
# meltpval = pvals
# meltpval$gene = rownames(meltpval)
# meltlogfc = logfcs
# meltlogfc$gene = rownames(meltlogfc)
# melted = merge(melt(meltpval),melt(meltlogfc), by = c("gene", "variable"), suffixes = c("pval","logfc"))
# colnames(melted) = gsub("value","",colnames(melted))
# colnames(melted)[2] = "Cluster"
# melted$Cluster = as.numeric(gsub("Cluster.","",melted$Cluster))
# 
# melted$pval[melted$pval==0] <- 10^-300
# 
# # topgenes = names(sort(apply(pvals, 1, min)))[1:100]
# # melted$gene = factor(melted$gene, levels = rev(names(sort(apply(pvals, 1, min)))))
# # 
# # ggplot(melted[melted$gene %in% topgenes,], 
# #        aes(x = Cluster, y = gene)) + geom_tile(aes(fill = -log10(pval)))
# 
# 
# topgenes = names(sort(apply(abs(logfcs), 1, max), decreasing = TRUE))[1:100]
# melted$gene = factor(melted$gene, levels = rev(names(sort(apply(abs(logfcs), 1, max), decreasing = TRUE))))
# 
# gmax = ggplot(melted[melted$gene %in% topgenes,], 
#        aes(x = Cluster, y = gene)) + geom_tile(aes(fill = logfc)) +
#   scale_fill_gradient2(high = scales::muted("red"), low = scales::muted("blue")) + 
#   labs(title = "Top 100 genes (max log2FC)")
# 
# #pheatmap(as.matrix(logfcs[topgenes,]), scale = "none")
# 
# topgenes = names(sort(abs(apply(logfcs, 1, mean)), decreasing = TRUE))[1:100]
# melted$gene = factor(melted$gene, levels = rev(names(sort(apply(abs(logfcs), 1, mean), decreasing = TRUE))))
# 
# gmean = ggplot(melted[melted$gene %in% topgenes,], 
#        aes(x = Cluster, y = gene)) + geom_tile(aes(fill = logfc)) +
#   scale_fill_gradient2(high = scales::muted("red"), low = scales::muted("blue")) + 
#   labs(title = "Top 100 genes (mean log2FC)")
# 
# plot_grid(gmax, gmean)

```



### Max 100 cells per group

To enable comparison between DE in different clusters, the DE analysis is redone using a smaller number of cells (100 per group). The results are shown in the same format as above.

```{r de_100, fig.height = 4, fig.width = 2.5}


if(!file.exists(paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_balanced.rds"))){
DGE_list_Treatment = list()

  for(i in sort(unique(alldata@active.ident))){
  message(paste("Cluster",i))
  # select all cells in cluster i
  cell_selection <- subset(alldata, cells = colnames(alldata)[alldata@active.ident == 
      i])
  cell_selection <- SetIdent(cell_selection, value = "Treatment")
  # Compute differential expression
  DGE_cell_selection <- FindMarkers(ident.1 = "GW", ident.2 = "STD",
                                    cell_selection, logfc.threshold = 0.2,
                                    test.use = "wilcox", max.cells.per.ident = 100,
      min.pct = 0.1, only.pos = FALSE,
      assay = "RNA")
  DGE_list_Treatment[[paste("Cluster",i)]] = DGE_cell_selection[DGE_cell_selection$p_val_adj<0.001,]
  DGE_list_Treatment[[paste("Cluster",i)]]$gene = rownames(DGE_list_Treatment[[paste("Cluster",i)]])
  saveRDS(DGE_list_Treatment, paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_balanced.rds"))
}}else{
  DGE_list_Treatment = readRDS(paste0("../analysis/DE_GWvsSTD_IMM_DN_",sel.lev,"_balanced.rds"))
}
```


```{r}

num.cells = as.matrix(table(alldata@active.ident,
                  alldata$Treatment))
num.cells[num.cells>100] = 100
rownames(num.cells) = paste0("Cluster ", rownames(num.cells))

reglist = lapply(names(DGE_list_Treatment), function(i){
  message(i)
  x = DGE_list_Treatment[[i]]
  c(up = sum(x$avg_log2FC > 0),
    down = sum(x$avg_log2FC < 0),
    total = nrow(x),
    min.cells = min(num.cells[i,]))})
names(reglist) = names(DGE_list_Treatment)
regdf = as.data.frame(reglist)
colnames(regdf) = gsub("Cluster.","", colnames(regdf))

```

Number of DE genes per cluster

`r knitr::kable(regdf)`


```{r,fig.height = 4, fig.width = 4}


tregdf = as.data.frame(t(regdf))
tregdf$Cluster = rownames(tregdf)
ggplot(tregdf, aes(x = as.numeric(Cluster), y = total)) + 
  geom_bar(stat = "identity")

```

```{r dotplots_100, fig.height = 4, fig.width = 2.5}

for(i in sort(unique(alldata@active.ident))){
  top20_cell_selection <- DGE_list_Treatment[[paste("Cluster",i)]]
  n = min(nrow(top20_cell_selection), 20)
  top20_cell_selection <- top20_cell_selection[order(top20_cell_selection$p_val)[1:n],]
  top20_cell_selection <- top20_cell_selection[order(top20_cell_selection$avg_log2FC)[1:n],]
  # print(VlnPlot(subset(alldata, idents = i),unique(top20_cell_selection$gene), 
  #         group.by = "orig.ident", assay = "RNA") + theme(axis.text = element_text(size = 1)))
  if(nrow(top20_cell_selection)>0){
    d = DotPlot(subset(alldata, idents = i), scale = FALSE, 
                features = rev(as.character(unique(top20_cell_selection$gene))), 
      assay = "RNA", group.by = "Treatment") + coord_flip() + plottheme +
        labs(title = paste("Cluster (",i,paste(top3$gene[top3$cluster==i],collapse = ", "),
                           ") (",nrow(DGE_list_Treatment[[paste("Cluster",i)]])," genes)"))
    d$guides$size$title = "Percent"
    d$guides$colour$title = "Average"
    print(d)
  }
}

```

```{r volcanos_100, fig.height = 4, fig.width = 7}


for(i in sort(unique(alldata@active.ident))){
  res = DGE_list_Treatment[[paste("Cluster",i)]]
  
  #res$avg_log2FC[res$cluster=="STD"] = 0-res$avg_log2FC[res$cluster=="STD"]
  
  if(nrow(res)>0){
      print(ggplot(res, aes(x = avg_log2FC, y = -log10(p_val_adj))) +
              geom_point(color = "#AAAAAA") + 
              geom_hline(yintercept = 0) +
              geom_vline(xintercept = 0) + 
              geom_text_repel(aes(label = gene), size = 2.5) +
              plottheme +
              theme(panel.background = element_blank(),
                    panel.grid = element_blank()) + 
        labs(title = paste("Cluster",i,paste(top3$gene[top3$cluster==i],collapse = ", "))))
  
  }
}



```



# GO inflammatory response


```{r, fig.width = 5}

Immune = c('Gal', 'Mefv', 'Ltb4r2', 'Hrh4', 'Irak2', 'Mir7116', 'Il17re', 'Ap3b1', 'Pja2',
           'Eif2ak1', 'Il17rc', 'Il17d', 'Gsdmd', 'Ttc39aos1', 'Orm2', 'Orm1', 'F12', 'Lat',
           'Nr1h4', 'Cxcl10', 'Cxcl9', 'Dab2ip', 'Ticam2', 'Tnip2', 'Nlrp1b', 'Tlr4', 'Ccl9',
           'F2rl1', 'Zfp580', 'Klk1b1', 'Ccl8', 'Il23r', 'Tcirg1', 'Il33', 'Nrros', 'Lacc1',
           'Tril', 'Itgb6', 'Calca', 'Cd180', 'Acvr1', 'Pla2g2e', 'Ackr2', 'S1pr3', 'Il1rl2',
           'Ccl17', 'Ccl20', 'Naip7', 'Ggt5', 'Tlr2', 'Mir301', 'Ltb4r1', 'Rps6ka4', 'Trem1',
           'Casp6', 'Tlr5', 'Myo5a', 'Csf1r', 'Csf1', 'Pik3cg', 'Il34', 'Tyrobp', 'Cxcr3',
           'Adcy1', 'Cxcl13', 'Serpinb1a', 'Acer3', 'Sema7a', 'Macir', 'Snca', 'Gpr33', 'Jam3',
           'Hdac7', 'Nupr1', 'Tollip', 'Ccl25', 'Alox5', 'Akt1', 'Ahcy', 'Agtr2', 'Agtr1b',
           'Agtr1a', 'Agt', 'Cxcl3', 'Adra2a', 'Nlrp6', 'Chst4', 'S100a7a', 'Il18r1', 'Il4ra',
           'Cxcr2', 'Rel', 'Camk1d', 'P2rx7', 'Cxcl15', 'Ndufs4', 'Epha2', 'Odam', 'Enpp1',
           'Notch2', 'Notch1', 'Nos2', 'Cxcl2', 'Stab1', 'Cd300a', 'Bmpr1b', 'Ahsg', 'Bcl6',
           'Ccr5', 'Serpinf2', 'Sigirr', 'Traf3ip2', 'Cx3cl1', 'Pycard', 'Irf5', 'Grn',
           'Nlrp4a', 'Rela', 'Il20rb', 'Ccl24', 'Zfp35', 'Il36rn', 'Il36a', 'Ddx3x', 'Ecm1',
           'Stat3', 'Tac4', 'Ccr7', 'Fut4', 'Fn1', 'Fcgr3', 'Tusc2', 'Fasn', 'Csrp3', 'Scyl1',
           'Ffar3', 'Cd6', 'Tnfaip3', 'Il17b', 'Scyl3', 'Chil1', 'Nlrp9a', 'Kng1', 'Cspg4',
           'Ackr1', 'C3ar1', 'Nlrp9b', 'Setd4', 'Mtor', 'Ulk4', 'Parp4', 'Chil3', 'Ptgir',
           'Syk', 'Ccl7', 'Ephx2', 'Ahcyl', 'Rarres2', 'Adora2a', 'Ext1', 'Acod1', 'Tbxa2r',
           'Loxl3', 'Gja1', 'B4galt1', 'Axl', 'Hyal3', 'Hdac9', 'Mecom', 'Tlr8', 'Tlr7', 
           'Tlr3', 'Tlr11', 'Tlr12', 'Tlr13', 'Cela1', 'Pla2g10', 'Nfkbiz', 'Sgms1', 'Pla2g7',
           'Slamf1', 'Hyal1', 'Chil4', 'Hp', 'Reg3b', 'Nlrp10', 'Rasgrp1', 'Hmox1', 'Hmgb2',
           'Hmgb1', 'Hk1', 'Tlr1', 'Jak2', 'Serpina1b', 'Cd47', 'Itgb2', 'Itgb1', 'Itgav',
           'Itgam', 'Trim55', 'Nlrc4', 'Il6', 'Il5ra', 'Il4', 'Il2ra', 'Il1rn', 'Il1r1', 
           'Casp1', 'Il1b', 'Il1a', 'Il13', 'Rbpj', 'Nlrp4e', 'Chia1', 'Themis2', 'Mif',
           'Clec10a', 'Mep1b', 'Mdk', 'Tpsb2', 'Nlrp4c', 'Pla2g2d', 'Trpv1', 'Afap1l2', 'Aim2',
           'Ptger4', 'Il23a', 'Lyn', 'Lrp1', 'Anxa1', 'Ffar4', 'Atrn', 'Tlr9', 'Adam8', 'Ccr4',
           'Casp4', 'Elane', 'Fut7', 'Ifng', 'Alox5ap', 'Ctnnbip1', 'Wdr83', 'Fpr1', 'Plgrkt',
           'Pnma1', 'Il17ra', 'Il17a', 'Igh-7', 'Ighg1', 'Ighg2a', 'Tlr6', 'Ido1', 'Nlrp4b',
           'Ncf1', 'Tnfrsf1a', 'Tnfrsf1b', 'Rtn4', 'Icam1', 'Il1f10', 'Ticam1', 'Adcy8', 'Hps1',
           'Serpina3n', 'Lipa', 'Il17c', 'Krt16', 'Kit', 'Jun', 'Il18', 'Zc3h12a', 'Hck', 'Hc',
           'Prkcz', 'Ifngr1', 'Ifngr2', 'Scn9a', 'Lxn', 'Pxk', 'Cd5l', 'Prkd1', 'Slc7a2', 
           'Chil6', 'Rap1gds1', 'Tnf', 'Lta', 'Trex1', 'Camp', 'Nfkbib', 'Ndst1', 'Scnn1b',
           'Cnr2', 'Ccr1', 'Ccr1l1', 'Ccr3', 'Slc11a1', 'Il31ra', 'Fpr-rs7', 'Fpr-rs6', 'Prcp',
           'Psen1', 'C5ar2', 'Mir21a', 'Mir155', 'Nampt', 'Dhx9', 'Seh1l', 'Il18rap', 'Lbp',
           'Il1rap', 'Nkiras2', 'Lncfao', 'Mfhas1', 'Cysltr1', 'Fpr3', 'Vnn1', 'Ccl26', 'Tnip1',
           'Ninj1', 'Ccl12', 'Ripk1', 'Pld4', 'Ak7', 'Kdm6b', 'Smo', 'Ly96', 'Ly86', 'Xcl1',
           'Pla2g3', 'Tnfrsf4', 'Tnfsf4', 'Pstpip1', 'P2rx1', 'Pik3cd', 'Adora3', 'Cxcl11',
           'Plaa', 'Slamf8', 'Ciita', 'Nfe2l2', 'Smpdl3b', 'Prkcq', 'Ccl21a', 'Ccl21b', 'Sdc1',
           'Chst2', 'Cxcl1', 'S100a9', 'Chil5', 'Il17f', 'C2cd4b', 'Myd88', 'Park7', 'Pbxip1',
           'H2bc1', 'Itgb2l', 'Elf3', 'Kl', 'Hpse', 'App', 'Stat5b', 'Selp', 'Sele', 'Cfh',
           'F8', 'F3', 'F2', 'Cebpb', 'Cd68', 'Cd44', 'Cd40lg', 'Cd40', 'Ptafr', 'Stard7',
           'Cd24a', 'Cd14', 'Aoc3', 'Ccl22', 'Umod', 'Casp7', 'Snap23', 'F7', 'Ano6', 'Ephb6',
           'Adam17', 'Ptgdr', 'Ccl11', 'Cd163', 'Cmklr1', 'Klkb1', 'Bdkrb2', 'Il36g', 'Itih4',
           'Ppbp', 'Relb', 'Nfkbid', 'Trp73', 'Tirap', 'Cxcl5', 'Msmp', 'Ccrl2', 'Nlrp1a',
           'Il25', 'C2cd4a', 'Nlrp4f', 'Mrgpra3', 'Ccl19', 'Gm5849', 'Ffar2', 'Stk39', 'Cers6',
           'Camk4', 'S100a8', 'C5ar1', 'C3', 'C1qa', 'Tac1', 'Bmp6', 'Bmp2', 'Bdkrb1',
           'Mapkapk2', 'Psen2', 'Rps6ka5', 'Unc13d', 'Ccl6', 'Ccl5', 'Ccl4', 'Ccl3', 'Ccl2',
           'Ccl1', 'Ager', 'Plscr1', 'Saa4', 'Saa3', 'Saa2', 'Saa1', 'Aif1', 'Pf4', 'Ms4a2',
           'Fcer1g', 'Fcgr1', 'Trem2', 'Gbp5', 'Cyba', 'Sphk1', 'Map2k3', 'Mapk8', 'Aimp1',
           'Nod2', 'Prdx2', 'Ywhaz', 'Wnk4', 'Smad1', 'Lilrb4a', 'Reg3a', 'Reg3g', 'F2r',
           'Naip2', 'Naip1', 'Naip6', 'Naip5', 'Chst1', 'Gper1', 'Cybb', 'Il36b', 'Nlrp9c',
           'Ccr6', 'Hdac5', 'Pld3', 'Serpinb9', 'Cyp26b1', 'Mylk3', 'Olr1', 'Dhrs7b', 'Fpr2',
           'Fpr-rs3', 'Fpr-rs4', 'Crlf2', 'Hif1a', 'Mir883b', 'Cd96', 'Crp', 'Timp1', 'Thbs1',
           'Tgfb1', 'Crhbp', 'Crh', 'Cr2', 'Ccr2', 'Havcr2', 'Cntf', 'Clu', 'Lias', 'Hnrnpa0',
           'Il27', 'Cxcr6', 'Cntnap2', 'Ptn', 'Tspan2', 'Ccl21c', 'Ptgs2', 'Ptgs1', 'Ptgfr',
           'Ptger3', 'Ptger2', 'Ptger1', 'Pparg', 'Polb', 'Cx3cr1', 'Nlrp3', 'Plp1', 'Dpep1')

alldata = AddModuleScore(alldata, features = list(Immune = Immune), name = "GOterm")

FeaturePlot(alldata, "GOterm1", min.cutoff = "q5", max.cutoff = "q95") +
  labs(title = "Inflammatory response") + NoAxes()

VlnPlot(alldata, "GOterm1", group.by = "orig.ident") + labs(title = "Inflammatory response")


```

